lib/metatron/composite_controller.rb

```diff
@@ -4,51 +4,65 @@ module Metatron
   # Implementes a Metacontroller CompositeController
   # @see https://metacontroller.github.io/metacontroller/api/compositecontroller.html
   class CompositeController < Controller
-    options "/sync" do
-      headers "Access-Control-Allow-Methods" => ["POST"]
-      halt 200
+    def initialize(env)
+      super
+      @strategy = nil
     end
 
-    post "/sync" do
-      if (provided_etag = calculate_sync_etag)
-        etag provided_etag
-      end
+    def calculate_customize_etag = nil
+    def calculate_sync_etag = nil
+    def customize = raise NotImplementedError
+    def finalize = raise NotImplementedError
+    def sync = raise NotImplementedError
 
-      data = sync
-      data[:children] = data[:children]&.map { |c| c.respond_to?(:render) ? c.render : c }
-      halt(data.to_json)
-    end
+    private
+
+    STRATEGY = {
+      "/customize" => { data: :customize, etag: :calculate_customize_etag },
+      # finalize calls should be rare and unique enough that we don't need to worry about ETags
+      "/finalize" => { data: :finalize },
+      "/sync" => { data: :sync, etag: :calculate_sync_etag }
+    }.freeze
+
+    def _call
+      return access_control_allow_methods if request.options?
+      return not_found unless request.post?
+
+      @strategy = STRATEGY.fetch(request.path_info) { return not_found }
+
+      headers = {}
 
-    options "/finalize" do
-      headers "Access-Control-Allow-Methods" => ["POST"]
-      halt 200
+      return Rack::Response[412, headers, []].to_a if fresh?(headers)
+
+      Rack::Response[200, headers, processed_data].to_a
     end
 
-    post "/finalize" do
-      # finalize calls should be rare and unique enough that we don't need to worry about ETags
+    def access_control_allow_methods
+      Rack::Response[200, { "access-control-allow-methods" => %w[POST] }, []].to_a
+    end
 
-      data = finalize
-      data[:children] = data[:children]&.map { |c| c.respond_to?(:render) ? c.render : c }
-      halt(data.to_json)
+    def not_found
+      Rack::Response[404, { "x-cascade" => "pass" }, []].to_a
     end
 
-    options "/customize" do
-      headers "Access-Control-Allow-Methods" => ["POST"]
-      halt 200
+    def fresh?(headers)
```

Comment:
Nit: `fresh?` feels a bit ambiguous to me. Maybe `request_still_fresh?` would be a little clearer?

Reply:
Sounds good to me, for full context here's what other frameworks do:

- Sinatra uses `etag_matches?` https://github.com/sinatra/sinatra/blob/734df89affd8de6350195bb7692848da2e3a7ec1/lib/sinatra/base.rb#L702
- Rails uses `fresh?`/`fresh_when` type terms
- Rack also has a `fresh?` method https://github.com/rack/rack/blob/f6c583adb0e863e524bacedaf594602964e01078/lib/rack/conditional_get.rb#L51

In our case (and Sinatra's) we aren't looking at `If-Modified-Since` at all so maybe the `fresh` terminology is less "correct" than something simple like `etag_matches?` 🤔